### PR TITLE
Fix staking on bech32

### PR DESCRIPTION
**Pull request by meyer9**

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -4065,7 +4066,8 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
 
     if (block.IsProofOfStake()) {
         // Coinbase output should be empty if proof-of-stake block
-        if (block.vtx[0].vout.size() != 1 || !block.vtx[0].vout[0].IsEmpty())
+        int commitpos = GetWitnessCommitmentIndex(block);
+        if (block.vtx[0].vout.size() != (commitpos == -1 ? 1 : 2) || !block.vtx[0].vout[0].IsEmpty())
```
Comment (meyer9) [inline src/main.cpp:4070]: Shouldn't `!block.vtx[0].vout[0].IsEmpty()` be changed to allow a segwit commitment?

Comment (wqking) [reply src/main.cpp:4070]: I think for segwit block.vtx[0].vout[0] should be empty and block.vtx[0].vout[1].IsEmpty() should not be empty?

File: src/main.cpp [hunk 2]
```diff
@@ -4376,11 +4372,11 @@ bool ContextualCheckBlock(const CBlock& block, CValidationState& state, CBlockIn
             // The malleation check is ignored; as the transaction tree itself
             // already does not permit it, it is impossible to trigger in the
             // witness tree.
-            if (block.vtx[1].wit.vtxinwit.size() != 1 || block.vtx[1].wit.vtxinwit[0].scriptWitness.stack.size() != 1 || block.vtx[1].wit.vtxinwit[0].scriptWitness.stack[0].size() != 32) {
+			if (block.vtx[0].wit.vtxinwit.size() != 1 || block.vtx[0].wit.vtxinwit[0].scriptWitness.stack.size() != 1 || block.vtx[0].wit.vtxinwit[0].scriptWitness.stack[0].size() != 32) {
```
Comment (meyer9) [inline src/main.cpp]: indentation needs to be fixed here

File: src/primitives/block.cpp [hunk 3]
```diff
@@ -207,12 +206,29 @@ bool CBlock::SignBlock(const CKeyStore& keystore)
             if (!keystore.GetKey(keyID, key))
                 return false;
 
-            //vector<unsigned char> vchSig;
             if (!key.Sign(GetHash(), vchBlockSig))
                  return false;
 
             return true;
         }
+        else if(whichType == TX_WITNESS_V0_SCRIPTHASH || whichType == TX_WITNESS_V0_KEYHASH)
```
Comment (meyer9) [inline src/primitives/block.cpp]: indentation here

Comment (meyer9) [reply src/primitives/block.cpp]: script hash addresses won't work. they require a larger script sig including the transaction data.

File: src/primitives/block.cpp [hunk 4]
```diff
@@ -259,6 +275,31 @@ bool CBlock::CheckBlockSignature() const
 
         return pubkey.Verify(GetHash(), vchBlockSig);
 
+    }
+    else if(whichType == TX_WITNESS_V0_SCRIPTHASH || whichType == TX_WITNESS_V0_KEYHASH)
```
Comment (meyer9) [inline src/primitives/block.cpp]: no SCRIPTHASH

File: src/wallet.cpp [hunk 5]
```diff
@@ -5416,4 +5416,4 @@ bool CWallet::DatabaseMint(CDeterministicMint& dMint)
     CWalletDB walletdb(strWalletFile);
     zphrTracker->Add(dMint, true);
     return true;
-}
```
Comment (meyer9) [inline src/wallet.cpp:5419]: unneeded change

**Review by meyer9 — COMMENTED** [review]
needs a bit cleaning up

File: src/primitives/block.cpp (hunk 3 above)
Comment (wqking) [reply src/primitives/block.cpp]: Fixed

**Comment by meyer9** [comment]
LGTM

**Comment by meyer9** [comment]
closed in favor of #126

**Pull request closed without merging**